blink/loader.c

```diff
@@ -471,20 +472,36 @@ static bool LoadElf(struct Machine *m,  //
   return execstack;
 }
 
-static void BootProgram(struct Machine *m,  //
-                        struct Elf *elf,    //
-                        void *map,          //
-                        size_t mapsize) {
-  m->cs.sel = m->cs.base = 0;
+void BootProgram(struct Machine *m,  //
+                 struct Elf *elf) {
+  int fd;
+  SetDefaultBiosIntVectors(m);
+  memset(m->beg, 0, sizeof(m->beg));  // reinitialize registers
+  memset(m->xmm, 0, sizeof(m->xmm));
+  memset(m->seg, 0, sizeof(m->seg));
   m->ip = 0x7c00;
   elf->base = 0x7c00;
-  memset(m->system->real, 0, 0x00f00000);
+  Write64(m->sp, 0x6f00);  // following QEMU
+  memset(m->system->real + 4 * 0x100, 0, kBiosBase - 4 * 0x100);
+  memset(m->system->real + 0x00100000, 0, kRealSize - 0x00100000);
   Write16(m->system->real + 0x400, 0x3F8);
   Write16(m->system->real + 0x40E, 0xb0000 >> 4);
+  Write16(m->system->real + 0x410, 1 << 0 |  // floppy drive
+                                   1 << 1 |  // math coprocessor
+                                   3 << 4 |  // initial video mode
+                                   0 << 6 |  // no. of floppy drives - 1
+                                   1 << 9);  // no. of serial devices
   Write16(m->system->real + 0x413, 0xb0000 / 1024);
   Write16(m->system->real + 0x44A, 80);
```

Comment:
Perhaps all these BDA (BIOS Data Area) initializations should move to `blink/defbios.c`, to keep BIOS-specific initializations out of the general `blink` INT 19h (or other) interrupt handlers? I like the idea of a separate defbios.c (even if only temporary until the ASM BIOS is implemented), as it keeps real mode BIOS initializations in a single place.

Reply:
@ghaerr : I have made the change.  Thanks!